File: 08_Exception_Handling/custom_exceptions.cpp
#include <iostream>
#include <stdexcept>
#include <string>

class InsufficientFundsException : public std::exception {
private:
    std::string m_message;
    double m_amountRequested;
    double m_currentBalance;

public:
    InsufficientFundsException(double requested, double balance) 
        : m_amountRequested(requested), m_currentBalance(balance) {
        m_message = "Insufficient Funds! Requested: $" + std::to_string(m_amountRequested) 
                    + ", Balance: $" + std::to_string(m_currentBalance);
    }

    const char* what() const noexcept override {
        return m_message.c_str();
    }

    double getRequestedAmount() const noexcept { return m_amountRequested; }
    double getCurrentBalance() const noexcept { return m_currentBalance; }
};

class BankAccount {
private:
    std::string m_owner;
    double m_balance;

public:
    BankAccount(const std::string& owner, double initialBalance) 
        : m_owner(owner), m_balance(initialBalance) {
        if (initialBalance < 0) {
            throw std::invalid_argument("Initial balance cannot be negative!");
        }
    }

    void deposit(double amount) {
        if (amount <= 0) {
            throw std::invalid_argument("Deposit must be positive!");
        }
        m_balance += amount;
    }

    void withdraw(double amount) {
        if (amount <= 0) {
            throw std::invalid_argument("Withdrawal must be positive!");
        }
        if (amount > m_balance) {
            throw InsufficientFundsException(amount, m_balance);
        }
        m_balance -= amount;
    }
};

int main() {
    try {
        BankAccount badAccount("Alice", -50.0);
    } 
    catch (const std::invalid_argument& e) {
        std::cout << "Caught expected error: " << e.what() << "\n";
    }

    try {
        BankAccount myAccount("Bob", 100.0);
        myAccount.withdraw(150.0);
    } 
    catch (const InsufficientFundsException& e) {
        std::cout << "Caught custom error: " << e.what() << "\n";
        std::cout << "  Requested: $" << e.getRequestedAmount() << "\n";
    }
    catch (const std::exception& e) {
        std::cout << "Caught error: " << e.what() << "\n";
    }

    return 0;
}

File: 08_Exception_Handling/java/CustomExceptions.java
// Custom exception classes, throws, and throw

public class CustomExceptions {

    static class InsufficientFundsException extends Exception {
        private final double deficit;

        InsufficientFundsException(double deficit) {
            super(String.format("Insufficient funds: need $%.2f more", deficit));
            this.deficit = deficit;
        }

        double getDeficit() { return deficit; }
    }

    static class InvalidAgeException extends IllegalArgumentException {
        InvalidAgeException(int age) {
            super("Invalid age: " + age + " (must be 0-150)");
        }
    }

    static class AccountLockedException extends RuntimeException {
        AccountLockedException(String accountId) {
            super("Account " + accountId + " is locked");
        }
    }

    static class BankAccount {
        private final String id;
        private double balance;
        private boolean locked;

        BankAccount(String id, double balance) {
            this.id = id;
            this.balance = balance;
            this.locked = false;
        }

        void deposit(double amount) throws IllegalArgumentException {
            if (amount <= 0) throw new IllegalArgumentException("Deposit must be positive");
            if (locked) throw new AccountLockedException(id);
            balance += amount;
            System.out.printf("  Deposited $%.2f -> Balance: $%.2f%n", amount, balance);
        }

        void withdraw(double amount) throws InsufficientFundsException {
            if (amount <= 0) throw new IllegalArgumentException("Withdrawal must be positive");
            if (locked) throw new AccountLockedException(id);
            if (amount > balance) throw new InsufficientFundsException(amount - balance);
            balance -= amount;
            System.out.printf("  Withdrew $%.2f -> Balance: $%.2f%n", amount, balance);
        }

        void lock() { locked = true; }
        double getBalance() { return balance; }
        String getId() { return id; }
    }

    static void validateAge(int age) throws InvalidAgeException {
        if (age < 0 || age > 150) throw new InvalidAgeException(age);
        System.out.println("  Age " + age + " is valid");
    }

    static void transfer(BankAccount from, BankAccount to, double amount)
            throws InsufficientFundsException {
        System.out.printf("Transferring $%.2f from %s to %s%n", amount, from.getId(), to.getId());
        from.withdraw(amount);
        to.deposit(amount);
        System.out.println("  Transfer complete!");
    }

    public static void main(String[] args) {
        System.out.println("===== Custom Checked Exception =====");
        BankAccount acc = new BankAccount("ACC-001", 100.0);
        try {
            acc.withdraw(50);
            acc.withdraw(75);
        } catch (InsufficientFundsException e) {
            System.out.println("Caught: " + e.getMessage());
            System.out.printf("Deficit: $%.2f%n", e.getDeficit());
        }

        System.out.println("\n===== Custom Unchecked Exception =====");
        int[] testAges = {25, -5, 200};
        for (int age : testAges) {
            try {
                validateAge(age);
            } catch (InvalidAgeException e) {
                System.out.println("  Caught: " + e.getMessage());
            }
        }

        System.out.println("\n===== RuntimeException (no catch required) =====");
        BankAccount locked = new BankAccount("ACC-002", 500.0);
        locked.lock();
        try {
            locked.deposit(100);
        } catch (AccountLockedException e) {
            System.out.println("Caught: " + e.getMessage());
        }

        System.out.println("\n===== Exception propagation with throws =====");
        BankAccount alice = new BankAccount("ALICE", 200.0);
        BankAccount bob = new BankAccount("BOB", 50.0);
        try {
            transfer(alice, bob, 100);
            transfer(bob, alice, 300);
        } catch (InsufficientFundsException e) {
            System.out.println("Transfer failed: " + e.getMessage());
        }

        System.out.println("\n===== Exception chaining =====");
        try {
            try {
                acc.withdraw(1000);
            } catch (InsufficientFundsException e) {
                throw new RuntimeException("Transaction failed", e);
            }
        } catch (RuntimeException e) {
            System.out.println("RuntimeException: " + e.getMessage());
            System.out.println("Caused by: " + e.getCause().getMessage());
        }

        System.out.printf("%nFinal balances: Alice=$%.2f, Bob=$%.2f%n",
                alice.getBalance(), bob.getBalance());
    }
}
